Check for authorization header.

sunnyrjuneja opened the pull request:
It appears that Swagger-UI correctly sets an authorization header when
added in the interface (api key field). However, if the endpoint has
the option "authorizations" set with "oauth2", it will override the
value. See https://github.com/ruby-grape/grape-swagger-rails/issues/13.

dblock commented on spec/dummy/app/api/api.rb:
```diff
@@ -26,5 +26,10 @@ class API < Grape::API
     request.params.as_json
   end
 
+  desc 'Get Authorization header.', authorizations: { oauth2: [] }
```
It would be nice if this actually did OAuth2 :) I would rename the endpoint either way to `oauth2`.

whatasunnyday replied:
I can set it up but it'll add the dep of doorkeeper (and activerecord). Okay, sounds good.

dblock replied:
I think a dep is fine, we want this to be closer to the real world.

dblock commented on spec/features/swagger_spec.rb:
```diff
@@ -49,8 +49,8 @@
       it 'adds an Authorization header' do
```
I'd like additional tests for this rather than editing existing tests. I am not sure how possible/easy it is because there's this global `api_auth` and `api_key_type` setup.

whatasunnyday replied:
I think it'll be pretty easy. Do you want 3 tests (1 for api key, 2 for header auth) or less?

dblock replied:
The more granular tests the better!

whatasunnyday replied:
It seems from the OAuth2 RFC that only bearer (and MAC but MAC is not widely deployed) is supported. See https://tools.ietf.org/html/rfc6750. I'll return the tests to the original state and add an additional testing making the focus OAuth2.

whatasunnyday commented:
@dblock Happy to improve these tests w/ feedback but how can we land a fix in Swagger UI? Do you know roughly where the offending code is?

Commit:
Check for oauth2 tokens.

It appears that Swagger-UI correctly sets an authorization header when
added in the interface (api key field). However, if the endpoint has
the option "authorizations" set with "oauth2", it will override the
value. See https://github.com/ruby-grape/grape-swagger-rails/issues/13.

whatasunnyday commented:
> It would be nice if this actually did OAuth2 :)

I ended up doing this but was forced to mock doorkeeper tokens which turned into a mess. Decided this was easiest way to mock. Anymore feedback @dblock?

dblock commented:
I think as far as tests go this is good.

whatasunnyday commented:
Cool, is the next move to open an issue at Swagger-UI?

dblock commented:
I think so. Supposedly once that's fixed there we can verify that this test passes.

supagroova commented:
@sunnyrjuneja did you ever manage to find a solution to this issue? I have the same problem from what I can tell and would like to investigate further to try and provide a solution.

sunnyrjuneja commented:
Hey @supagroova excuse my brevity since I'm on my phone. There are some references PRs in repos above. I essentially had to do a few work around but was able to solve my problem. If you're unable to figure it out, I'm happy to chat on Skype or GHangout for your problem.

matfiz commented:
Hey @sunnyrjuneja @supagroova! I have somehow managed to get it up and running. What I did was I have changed the way the `GrapeSwaggerRails.options.headers` are passed to `SwaggerUi`. 
The problem in the original implementation is that the headers are read into `swaggerUi.api.clientAuthorizations` after the `SwaggerUi` is loaded. It means upon the request to Swagger JSON, the headers are not set!

The solution I have found is that instead of reading it after `SwaggerUi.load()`, I enter them directly to the `SwaggerUi` initializer (look how authorizations are loaded in https://github.com/swagger-api/swagger-ui/blob/master/src/main/javascript/SwaggerUi.js#L100).

The working code after changes is:
```javascript
    auth_headers = {}
    <% GrapeSwaggerRails.options.headers.each_with_index do |(key, value), index| %>
      <%=raw "auth_headers['header_#{index}'] = new SwaggerClient.ApiKeyAuthorization('#{CGI.escapeHTML(key)}', '#{CGI.escapeHTML(value)}', 'header');" %>
    <% end %>

    window.swaggerUi = new SwaggerUi({
      url: options.app_url + options.url,
      dom_id: "swagger-ui-container",
      supportHeaderParams: true,
      supportedSubmitMethods: ['get', 'post', 'put', 'delete', 'patch'],
      onComplete: function(swaggerApi, swaggerUi){
        if('console' in window) {
          console.log("Loaded SwaggerUI")
          console.log(swaggerApi);
          console.log(swaggerUi);
        }
        $('pre code').each(function(i, e) {hljs.highlightBlock(e)});
      },
      onFailure: function(data) {
        if('console' in window) {
          console.log("Unable to Load SwaggerUI");
          console.log(data);
        }
      },
      docExpansion: options.doc_expansion,
      apisSorter: "alpha",
      authorizations: auth_headers
    });

You can see the full new `index.html.erb` here: https://github.com/GeecoLABs/grape-swagger-rails/blob/master/app/views/grape_swagger_rails/application/index.html.erb. 

If you find it of any interest @dblock, I can prepare a separate PR.

sunnyrjuneja commented:
@matfiz That looks great. Thanks mat!

supagroova commented:
@matfiz nice, I like it! I ended up using the latest swagger-ui version and doing to following based on the code found here - https://github.com/swagger-api/swagger-ui/blob/master/dist/index.html#L81

Then I changed it as such: 
```javascript
      function addApiKeyAuthorization(){
        var key = encodeURIComponent($('#input_apiKey')[0].value);
        if(key && key.trim() != "") {
            apiKeyAuth = new SwaggerClient.ApiKeyAuthorization("Authorization", "Bearer "+key, "header");
            window.swaggerUi.api.clientAuthorizations.add("oauth2", apiKeyAuth);
            log("added key " + key);
        }
      }
```

The key difference is that once I changed the key `api_key` to `oauth2` the headers were then included in the subsequent ajax calls.

I like you're approach too and will try it out as I'd like to use this gem and keep swagger-ui out of my app.

supagroova commented:
@matfiz looking at your code again it would seem to me that the `api_key` you're setting in you're headers must be defined at app initialisation no?

My approach was to let the end user input their api_key into the `#input_apiKey` input and have that applied to all ajax calls.

I'm thinking I might fork the project, update it to the latest version of swagger-ui and then add this functionality. Might be of interest to you guys?

matfiz commented:
@supagroova Pls have a look at all the changes I made to `index.html.erb` https://github.com/GeecoLABs/grape-swagger-rails/blob/master/app/views/grape_swagger_rails/application/index.html.erb

In my solution, the `api_key` can be initialized either at app initialization or after entering `#input_apiKey`.

What you suggested: `window.swaggerUi.api.clientAuthorizations.add("oauth2", apiKeyAuth);` with changing `api-key` to `oauth2` actually should not affect anything, as this just changes the label in `authz`. 

I think I was solving a problem a little bit different that yours. I needed the header to be included in the api call to fetch `swagger_doc.json`.  `window.swaggerUi.api` is not availale before `window.swaggerUi.load()`, so the `clientAuthorizations.add()` cannot be used for that. That's why I've included the manualy constructed `authz` in the `swaggerUi` init options. 

I hope it has clarified :) I am also happy to provide a clean PR with the changes I've described.

dblock commented:
I haven't dug in the discussion above, but I am interested in fixing anything that's broken here, please make pull requests!

davidbrewer commented:
@supagroova You saved my life with your `addApiKeyAuthorization` example above. Could not for the life of me get api keys via Authorization headers working until I followed your example. Thanks!

@dblock: I'm not sure how best to make a clean pull request out of this change... there are still some weird issues in the resulting behavior (such as some javascript errors that remain, and the curl field not getting updated), but I think those weird issues might already be present and be related to multipart forms rather than header-based authorization. I suspect an update to the underlying Swagger-UI code might fix some of these issues but I'm a bit out of my depth when it comes to the other consequences.

Can you take a look at this commit and tell me if it seems worth trying to make a pull request out of this?

https://github.com/davidbrewer/grape-swagger-rails/commit/5701ef57241eac18fbd81444a3e97d0375323586

Apologies if this has all gotten a bit far afield from the original purpose of this pull request!

supagroova commented:
@davidbrewer It took me quite some time to work out that fix, so I'm glad that helped you as well! :-)

dblock commented:
I commented in #48.

dblock commented:
I believe https://github.com/ruby-grape/grape-swagger-rails/pull/48 fixes this, I am going to close it.

dblock commented:
It still fails:

```
  1) Swagger #options #api_auth:oauth2 adds a token when the route specifies oauth2 authorization
     Failure/Error: expect(page).to have_css 'span.attribute', text: 'Authorization'
       expected to find css "span.attribute" with text "Authorization" but there were no matches
     # ./spec/features/swagger_spec.rb:108:in `block (4 levels) in <top (required)>'
```

Want to take a look with all the recent changes @sunnyrjuneja or @davidbrewer?

sunnyrjuneja commented:
I can take a look this afternoon. I'll block out sometime for 1pm PST.

dblock commented:
Given all the 💰 we pay you you're now totally on a deadline :)

sunnyrjuneja commented:
Hey @dblock I gave a pretty honest look at this. I think I can help but do you mind explaining how to run the test server without running tests?

I tried `bundle exec rackup config.ru` in grape-swagger-rails/spec/dummy but I get Can't read from server. It may not have the appropriate access-control-origin settings.

dblock commented:
You're on the right track, it's cause the default is pointing to the wrong place. Change the URL in the green box to http://localhost:9292.

<img width="1000" alt="screen shot 2016-04-28 at 9 19 18 am" src="https://cloud.githubusercontent.com/assets/542335/14887124/576b93ba-0d22-11e6-8d5e-038869f4deca.png">

moskvin commented:
Thanks. The concern here may still be valid, but this PR is too old to merge as-is.
The reasons:
- it targets the old Swagger UI interaction model
- it only adds tests, and those tests are written against an outdated UI structure/flow
- the project has since moved to a different frontend/runtime architecture

If this bug still exists on current master, the next step would be:
- reproduce it against the current Swagger UI v5 integration
- rewrite the regression spec against the current test style and current UI/runtime
- add a fix only if the bug is still present

This PR should be closed. Please open a fresh PR with a modern failing test and fix.
